vm/src/obj/objset.rs

```diff
@@ -24,21 +25,80 @@ pub struct PySet {
 }
 pub type PySetRef = PyRef<PySet>;
 
+#[derive(Default)]
+pub struct PyFrozenSet {
+    elements: HashMap<u64, PyObjectRef>,
+}
+pub type PyFrozenSetRef = PyRef<PyFrozenSet>;
+
 impl fmt::Debug for PySet {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         // TODO: implement more detailed, non-recursive Debug formatter
         f.write_str("set")
     }
 }
 
+impl fmt::Debug for PyFrozenSet {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // TODO: implement more detailed, non-recursive Debug formatter
+        f.write_str("frozenset")
+    }
+}
+
 impl PyValue for PySet {
     fn class(vm: &VirtualMachine) -> PyClassRef {
         vm.ctx.set_type()
     }
 }
 
+impl PyValue for PyFrozenSet {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.ctx.frozenset_type()
+    }
+}
+
 pub fn get_elements(obj: &PyObjectRef) -> HashMap<u64, PyObjectRef> {
-    obj.payload::<PySet>().unwrap().elements.borrow().clone()
+    if let Some(set) = obj.payload::<PySet>() {
+        return set.elements.borrow().clone();
+    } else if let Some(frozenset) = obj.payload::<PyFrozenSet>() {
+        return frozenset.elements.clone();
+    }
+    panic!("Not frozenset or set");
+}
+
+macro_rules! validate_set_or_frozenset {
```

Comment:
I don't think this needs to be a macro. This could be defined as:
```rust
fn validate_set_or_frozenset(vm: &VirtualMachine, obj: &PyObjectRef) -> PyResult<()> {
     ...
}
```
And use ? to propagate the error.

Reply:
Done

Reply:
Agreed, although I am in favor of macros in some places, functions are usually preferred. Good review!